---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
library(metacoder)
library(tidyverse)
library(ggpubr)
library(ggtext)
library(patchwork)
library(vegan)
library(ggnewscale)
library(ggtree)
source(here::here("code/funcs.R"))

set.seed(2022)
conflicted::conflicts_prefer(dplyr::select(), dplyr::rename(), dplyr::filter())
```

# Community stats

## Load data
```{r}
metadata <- read_metaG_metadata()
```
Previously used this bracken data, but have since replace with relative abundance from bins
```{r}
rel.abund_bracken <- read_tsv("data/sample_data/bracken_rel_abund.tsv")

vegan_abund_bracken <- rel.abund_bracken %>% 
  dplyr::select(taxonomy, any_of(metadata$import_id)) %>% 
  column_to_rownames("taxonomy") %>% 
  t()

rel_abund_long_bracken <- rel.abund_bracken %>% 
  pivot_longer(`0245e67227092c6391e231e929be7edd`:ncol(.), names_to = "import_id", values_to = "rel_abund")
```

To load from bin abundance instead:
```{r}

percent_abund <- read_rds("results/percent_abund.rds")

rel.abund <- percent_abund %>% 
  dplyr::select(taxonomy = "classification", Domain, Phylum, Class, Order, Family, Genus, Species, taxonomy_id = "genome", percent_abund, sample) %>% 
  pivot_wider(values_from = "percent_abund", names_from = "sample",values_fill = 0) %>% 
  mutate(taxonomy = str_remove_all(taxonomy,"(;[a-z]__)+$"),
         taxonomy = if_else(taxonomy == "Unclassified Bacteria", "d__Bacteria", taxonomy),
         taxonomy = if_else(taxonomy == "Unclassified" | is.na(taxonomy) , "", taxonomy),
         taxonomy = paste0("r__root;",taxonomy),
         taxonomy = str_remove(taxonomy, ";$"))
  
rel_abund_long <- rel.abund %>% 
  pivot_longer(any_of(metadata$sample), names_to = "sample",values_to = "rel_abund")

vegan_abund <- percent_abund %>% 
  ungroup() %>% 
  dplyr::select(genome,sample, percent_abund) %>% 
  distinct() %>% 
  pivot_wider(names_from = sample, values_from = percent_abund, values_fill = 0) %>% 
  column_to_rownames("genome") %>% 
  t()

vegan_abund_family <- percent_abund %>% 
  ungroup() %>% 
    group_by(Phylum, Class, Order, Family, sample) %>% 
    summarise(percent_abund = sum(percent_abund),
              lineage = paste(Phylum, Class, Order, Family, sep = "; ")) %>% 
  distinct() %>% ungroup() %>% 
  dplyr::select(sample, lineage, percent_abund) %>% 
  pivot_wider(names_from = sample, values_from = percent_abund, values_fill = 0) %>% 
  column_to_rownames("lineage") %>% 
  t()

```

## Approx microbial biomass
```{r}

scaling_factor_from_spike_ins <-  read_rds("results/scaled_percent_abund.rds") %>% dplyr::select(sample, scaling_factor)

read_count_paths <- system("ls data/omics/metagenomes/*/reads/*_read_count_fastp.tsv", intern = TRUE) %>% 
  data.frame(path = .) %>% 
  cbind(., .$path %>% unglue::unglue_data("data/omics/metagenomes/{sample}/reads/{samp_name}_read_count_fastp.tsv"))

import_read_counts <- function(path, sample_id){
  read_tsv(path,show_col_types = FALSE ) %>% 
    mutate(sample = sample_id)
}

read_counts <- map2_dfr(read_count_paths$path, read_count_paths$sample, import_read_counts)

read_count_stats <- read_counts %>% 
  dplyr::select(sample, fwd_read_count, read_state) %>% 
  pivot_wider(names_from = read_state, values_from = fwd_read_count) %>% 
  mutate(percent_of_reads_lost_to_decontam =  (filt_and_trimmed_reads - decon_reads) / filt_and_trimmed_reads) %>% 
  left_join(metadata)

read_count_stats %>% 
  ggplot(aes(gut_section, percent_of_reads_lost_to_decontam, color = transplant_type)) +
  geom_boxplot() +
  theme_bw()

read_count_stats %>% 
  ggplot(aes(gut_section, 1- percent_of_reads_lost_to_decontam, color = transplant_type)) +
  geom_boxplot() +
  theme_bw()

# Strong agreement between reads mapping to mouse genome and scaling factor determined from spike-ins
read_count_stats %>% 
  left_join(scaling_factor_from_spike_ins) %>% 
  ggplot(aes(1 - percent_of_reads_lost_to_decontam, scaling_factor)) +
  geom_point(aes(shape = gut_section, color = plot_categories)) +
  geom_smooth(method = "lm") +
  theme_bw() +
  labs(y = "Scaling factor determined from spike-ins",
       x = "Percent of reads not mapping to mouse genome")
ggsave("results/abundance_normalization.pdf",width = 3, height = 2, scale = 1.75)

```


```{r}
spike_in_conc <- read_excel("data/sample_data/spike_in_concs.xlsx") %>% 
  left_join(metadata) %>% 
  pivot_longer(any_of(ends_with("genome_copies")),names_to = "genome", values_to = "genome_copies") %>% 
  left_join(read_count_stats %>% dplyr::select(researcher_sample_id, percent_of_reads_lost_to_decontam))



spike_in_conc %>% 
  ggplot(aes(1-percent_of_reads_lost_to_decontam, genome_copies, color = gut_section)) +
  geom_point()


```





## Diversity

### Alpha diversity

#### Shannon

##### Day 21 only
```{r}
shannon <- vegan::diversity(vegan_abund) %>% 
  data.frame(import_id = names(.), shannon = .) %>% 
  left_join(metadata) 

shannon %>% 
  filter(days_after_transplant == 21) %>% 
  ggplot(aes(gut_section, shannon, color = plot_categories)) +
  geom_boxplot() +
  geom_point(position = position_jitterdodge()) +
  scale_color_manual(values = c("#2083e6", "#f09f26")) +
  labs(y= "Shannon diversity", color = NULL, x = NULL) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -25, hjust = 0, vjust = 1))

ggsave("results/shannon_diversity_day21.pdf", width = 5, height = 3, dpi = 600, scale = 1)
ggsave("results/shannon_diversity_day21.png", width = 5, height = 3, dpi = 600, scale = 1)

(shan_stat_plot21 <- shannon %>% 
  filter(days_after_transplant == 21) %>% 
  ggboxplot(x = "gut_section", y = "shannon",
          color = "plot_categories", palette = "jco",
          add = "jitter") +
    scale_color_manual(values = c("#2083e6", "#f09f26")) +
  stat_compare_means(aes(group = plot_categories),method = "t.test") + # label = "p.signif" 
    labs(x = NULL, y = "Shannon diversity", color = NULL) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    expand_limits(x = c(1,5))
)

ggsave("results/shannon_diversity_stats_day21.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/shannon_diversity_stats_day21.png", width = 5, height = 3, dpi = 600, scale = 1.5)
```

##### Full
```{r}
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

shannon %>% 
  ggplot(aes(gut_section, shannon, color = plot_categories)) +
  geom_boxplot() +
  geom_point(position = position_jitterdodge()) +
  scale_color_manual(values = colors) +
  labs(y= "Shannon diversity", color = NULL, x = NULL) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -25, hjust = 0, vjust = 1))

ggsave("results/shannon_diversity.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/shannon_diversity.png", width = 5, height = 3, dpi = 600, scale = 1.5)

(shan_stat_plot <- shannon %>% 
  ggboxplot(x = "gut_section", y = "shannon",
          color = "plot_categories", palette = "jco",
          add = "jitter") +
    scale_color_manual(values = colors) +
  stat_compare_means(aes(group = plot_categories),method = "t.test") + # label = "p.signif"
    labs(x = NULL, y = "Shannon diversity", color = NULL) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    expand_limits(x = c(1,5))
)

ggsave("results/shannon_diversity_stats.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/shannon_diversity_stats.png", width = 5, height = 3, dpi = 600, scale = 1.5)


for_model <- vegan::diversity(vegan_abund,index = "shannon") %>% 
  data.frame(import_id = names(.), invsimpson = .) %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% c("Allogeneic", "Syngeneic")) %>% 
  mutate(transplant_type = relevel(factor(transplant_type), "Syngeneic"),
         gut_section = factor(gut_section, ordered = FALSE),
         gut_section = relevel(factor(gut_section), "Transverse Colon"),
         days_after_transplant = factor(days_after_transplant, ordered = FALSE)) 

shannon_mod <- lmerTest::lmer(invsimpson ~ gut_section + transplant_type * days_after_transplant + (1|replicate), data = for_model)

summary(shannon_mod)

```

#### Inverse Simpson
##### Day 21 only
```{r}
inv_simpson <- vegan::diversity(vegan_abund,index = "invsimpson") %>% 
  data.frame(import_id = names(.), invsimpson = .) %>% 
  left_join(metadata) %>% 
  mutate(transplant_type = relevel(factor(transplant_type), "Naive"))

inv_simpson %>% 
  filter(days_after_transplant == 21) %>% 
  ggplot(aes(gut_section, invsimpson, color = plot_categories)) +
  geom_boxplot() +
  geom_point(position = position_jitterdodge()) +
  scale_color_manual(values = c("mediumpurple3", "olivedrab")) +
  #scale_color_manual(values = colors) +
  labs(y= "Inverse Simpson index", color = NULL, x = NULL) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -25, hjust = 0, vjust = 1))

ggsave("results/invsimpson_diversity_day21.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/invsimpson_diversity_day21.png", width = 5, height = 3, dpi = 600, scale = 1.5)


(simp_stat_plot21 <- inv_simpson %>% 
  filter(days_after_transplant == 21) %>% 
  ggboxplot(x = "gut_section", y = "invsimpson",
          color = "plot_categories", palette = "jco",
          add = "jitter") +
  stat_compare_means(aes(group = plot_categories),method = "t.test") +
    labs(x = NULL, y = "Inverse Simpson index", color = NULL) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    theme(legend.position = "none",
          axis.text.x = element_blank()) +
    expand_limits(x = c(1,5))
)

ggsave("results/invsimpson_diversity_stats_day21.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/invsimpson_diversity_stats_day21.png", width = 5, height = 3, dpi = 600, scale = 1.5)

```

##### Full
```{r}
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

inv_simpson %>% 
  #filter(days_after_transplant == 21) %>% 
  ggplot(aes(gut_section, invsimpson, color = plot_categories)) +
  geom_boxplot() +
  geom_point(position = position_jitterdodge()) +
  #scale_color_manual(values = c("mediumpurple3", "olivedrab")) +
  scale_color_manual(values = colors) +
  labs(y= "Inverse Simpson index", color = NULL, x = NULL) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -25, hjust = 0, vjust = 1))

ggsave("results/invsimpson_diversity.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/invsimpson_diversity.png", width = 5, height = 3, dpi = 600, scale = 1.5)


(simp_stat_plot <- inv_simpson %>% 
  filter(days_after_transplant == 21) %>% 
  ggboxplot(x = "gut_section", y = "invsimpson",
          color = "plot_categories", palette = "jco",
          add = "jitter") +
  stat_compare_means(aes(group = plot_categories),method = "t.test") +
    labs(x = NULL, y = "Inverse Simpson index", color = NULL) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    theme(legend.position = "none",
          axis.text.x = element_blank()) +
    expand_limits(x = c(1,5))
)

ggsave("results/invsimpson_diversity_stats.pdf", width = 5, height = 3, dpi = 600, scale = 1.5)
ggsave("results/invsimpson_diversity_stats.png", width = 5, height = 3, dpi = 600, scale = 1.5)

simp_stat_plot / shan_stat_plot + theme(legend.position = "bottom")
ggsave("results/diversity_day_21.pdf", width = 3, height = 4, scale = 2)
ggsave("results/diversity_day_21.png", width = 3, height = 4, scale = 2)


for_model <- vegan::diversity(vegan_abund,index = "invsimpson") %>% 
  data.frame(import_id = names(.), invsimpson = .) %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% c("Allogeneic", "Syngeneic")) %>% 
  mutate(transplant_type = relevel(factor(transplant_type), "Syngeneic"),
         gut_section = factor(gut_section, ordered = FALSE),
         gut_section = relevel(factor(gut_section), "Transverse Colon"),
         days_after_transplant = factor(days_after_transplant, ordered = FALSE)) 

invSimp_mod <- lmerTest::lmer(invsimpson ~ gut_section + transplant_type * days_after_transplant + (1|replicate), data = for_model)

summary(invSimp_mod)

```



### Plot both Shannon and Simpson diversity together

```{r}

combined_alpha <- shannon %>%
  rename(diversity = "shannon") %>% 
  mutate(method = "Shannon") %>% 
  bind_rows(inv_simpson %>% rename(diversity ="invsimpson") %>% mutate(method = "Inverse Simpson"))

combined_alpha %>% 
  ggplot(aes(gut_section, diversity, color = plot_categories)) +
  geom_boxplot() +
  geom_point(position = position_jitterdodge()) +
  #scale_color_manual(values = c("mediumpurple3", "olivedrab")) +
  scale_color_manual(values = colors) +
  labs(y= "Alpha (within sample) diversity", color = NULL, x = NULL) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = -25, hjust = 0, vjust = 1),
        legend.position = "top") +
  facet_wrap(~method,scales = "free_y")


colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

comparisons <- list(c("Allogeneic; day 7", "Allogeneic; day 21"), c("Syngeneic; day 7", "Syngeneic; day 21"))

comparisons <- list(c("#f09f26", "#dec8a6"), c("#9dbfe0", "#2083e6"))


comparisons <- list(c("Descending Colon", "Transverse Colon"), c("Terminal ileum", "Cecum"))


comparisons <- list(c(2,3), c(4,5),c(2,4),c(3,5) )

(combined_simp_stat_plot <- combined_alpha %>% 
  ggboxplot(x = "plot_categories", y = "diversity",
          color = "plot_categories", palette = "jco",
          add = "jitter") +
  stat_compare_means(comparisons = comparisons,method = "t.test") +
    labs(x = NULL, y = "Alpha (within sample) diveristy", color = NULL) +
    scale_x_discrete(guide = guide_axis(angle = -45)) +
    scale_y_continuous(expand = expansion(mult = c(0,.1))) +
    scale_color_manual(values = colors) +
    facet_grid(method ~ gut_section,scales = "free_y") +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
)

ggsave("results/alpha_diversity.png",width = 2.75, height = 1.75, scale = 2.75)
ggsave("results/alpha_diversity.pdf",width = 2.75, height = 1.75, scale = 2.75)

```



## Beta diversity
### PERMANOVA testing
```{r}
abund_mat <- rel.abund %>% 
  column_to_rownames("taxonomy_id") %>% 
  dplyr::select(any_of(metadata %>% filter(!is.na(days_after_transplant)) %>% pull(import_id))) %>% 
  t()

tax_data <- rel.abund %>% 
  dplyr::select(taxonomy_id, taxonomy)

vegan_met <- metadata %>% 
  column_to_rownames("import_id") %>% 
  .[rownames(abund_mat),]

(permanova_res_seqeuntial <- vegan::adonis2(abund_mat ~ gut_section + transplant_type * days_after_transplant,data = vegan_met))


(permanova_res_marginal <- vegan::adonis2(abund_mat ~ gut_section + transplant_type * days_after_transplant,data = vegan_met,by = "margin"))
```


### NMDS
#### All days
```{r}
NMDS <- vegan::metaMDS(vegdist(vegan_abund,method = "bray"))

NMDS

NMDS_points <- NMDS$points %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata)

#tax_fit <- envfit(NMDS, vegan_abund, permutations = 100, na.rm = TRUE)

colors = c("#9a5ebd","#2083e6","#2083e6", "#f09f26","#f09f26")
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

(NMDS_points %>% 
  mutate(type_and_days = paste0(transplant_type, "_", days_after_transplant),
         location_type_and_days = paste0(gut_section, "_", type_and_days)) %>% 
  ggplot(aes(MDS1, MDS2, color = plot_categories, shape = gut_section)) +
  scale_alpha_discrete(range = c(0.4,1)) +
  geom_point(aes(group = plot_categories)) +
  scale_color_manual(values = colors) +
  new_scale_color() +
  ggforce::geom_mark_ellipse(data = . %>% filter(days_after_transplant != "other"),aes(group = days_after_transplant, label = str_glue("day {days_after_transplant}")),label.fontsize = 8, color = "grey70") +
  #coord_fixed(xlim = c(-150, 100), ylim = c(-100, 60)) +
  labs(color = NULL, shape = NULL) +
  theme_bw() ) #%>% plotly::ggplotly()

ggsave("results/community_NMDS.pdf", width = 6, height = 3, dpi = 600, scale = 1.25)
ggsave("results/community_NMDS.png", width = 6, height = 3, dpi = 600, scale = 1.25)

```

#### All days, no terminal Illeum
```{r}

no_ti <- metadata %>% 
  filter(gut_section != "Terminal ileum")

veg_abund_no_TI <- vegan_abund[no_ti$import_id,]

NMDS <- vegan::metaMDS(vegdist(veg_abund_no_TI,method = "bray"))

NMDS

NMDS_points <- NMDS$points %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata)

#tax_fit <- envfit(NMDS, vegan_abund, permutations = 100, na.rm = TRUE)

colors = c("#9a5ebd","#2083e6","#2083e6", "#f09f26","#f09f26")
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

(NMDS_points %>% 
  mutate(type_and_days = paste0(transplant_type, "_", days_after_transplant),
         location_type_and_days = paste0(gut_section, "_", type_and_days)) %>% 
  ggplot(aes(MDS1, MDS2, color = plot_categories, shape = gut_section)) +
  scale_alpha_discrete(range = c(0.4,1)) +
  geom_point(aes(group = plot_categories)) +
  scale_color_manual(values = colors) +
  new_scale_color() +
  ggforce::geom_mark_ellipse(data = . %>% filter(days_after_transplant != "other"),aes(group = days_after_transplant, label = str_glue("day {days_after_transplant}")),label.fontsize = 8, color = "grey70") +
  #coord_fixed(xlim = c(-75, 125), ylim = c(-80, 60)) +
  labs(color = NULL, shape = NULL) +
  theme_bw() ) #%>% plotly::ggplotly()

ggsave("results/community_NMDS_noTI.pdf", width = 6, height = 3, dpi = 600, scale = 1.25)
ggsave("results/community_NMDS_noTI.png", width = 6, height = 3, dpi = 600, scale = 1.25)

```



#### Day 21 only

```{r}

day21_met <- metadata %>% 
  filter(days_after_transplant == 21)

veg_abund_day21 <- vegan_abund[day21_met$import_id,]

NMDS <- vegan::metaMDS(vegdist(veg_abund_day21,method = "robust.aitchison"))

NMDS

NMDS_points <- NMDS$points %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata)

#tax_fit <- envfit(NMDS, vegan_abund, permutations = 100, na.rm = TRUE)

colors = c("#9a5ebd","#2083e6","#2083e6", "#f09f26","#f09f26")
colors = c("#2083e6","#f09f26")

(NMDS_points %>% 
  mutate(type_and_days = paste0(transplant_type, "_", days_after_transplant),
         location_type_and_days = paste0(gut_section, "_", type_and_days)) %>% 
  ggplot(aes(MDS1, MDS2, color = plot_categories, shape = gut_section)) +
  scale_alpha_discrete(range = c(0.4,1)) +
  geom_point(aes(group = plot_categories)) +
  #ggforce::geom_mark_ellipse(data = . %>% filter(days_after_transplant != "other"),aes(group = days_after_transplant, label = str_glue("day {days_after_transplant}")),label.fontsize = 10) +
    ggforce::geom_mark_ellipse(aes(group = plot_categories),label.fontsize = 10) +
  scale_color_manual(values = colors) +
  coord_fixed(xlim = c(-80, 65), ylim = c(-60, 70)) +
  labs(color = NULL, shape = NULL) +
  theme_bw() ) #%>% plotly::ggplotly()

ggsave("results/community_NMDS_day21.pdf", width = 4.5, height = 3, dpi = 600, scale = 1.25)
ggsave("results/community_NMDS_day21.png", width = 4.5, height = 3, dpi = 600, scale = 1.25)

```


#### With envfit CCA subset of microbes & metabolites

```{r}
no_ti <- metadata %>% 
  filter(gut_section != "Terminal ileum")

veg_abund_no_TI <- vegan_abund[no_ti$import_id,]

CCA_res <- read_rds("results/CCA_multiomics_community_metabolite_simplified.rds")

NMDS <- vegan::metaMDS(vegdist(veg_abund_no_TI[rownames(CCA_res),],method = "bray"))

en <- envfit(NMDS, CCA_res, permutations = 100, na.rm = TRUE)

env_res_df <- env_res$vectors$arrows



NMDS_points <- NMDS$points %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata)

#tax_fit <- envfit(NMDS, vegan_abund, permutations = 100, na.rm = TRUE)

colors = c("#9a5ebd","#2083e6","#2083e6", "#f09f26","#f09f26")
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

(NMDS_points %>% 
  mutate(type_and_days = paste0(transplant_type, "_", days_after_transplant),
         location_type_and_days = paste0(gut_section, "_", type_and_days)) %>% 
  ggplot(aes(MDS1, MDS2, color = plot_categories, shape = gut_section)) +
  scale_alpha_discrete(range = c(0.4,1)) +
  geom_point(aes(group = plot_categories)) +
  ggforce::geom_mark_ellipse(aes(group = days_after_transplant, label = str_glue("{days_after_transplant} days")),label.fontsize = 10) +
  scale_color_manual(values = colors) +
  coord_fixed(xlim = c(-150, 100), ylim = c(-100, 60)) +
  labs(color = NULL, shape = NULL) +
  theme_bw() ) #%>% plotly::ggplotly()

bin_taxonomy <- rel.abund %>% 
  dplyr::select(taxonomy, Species = "taxonomy_id")


spp.pval <- en$vectors$pvals %>% data.frame(p.value = .)
spp.scrs <- as.data.frame(scores(en, display = "vectors")) 
spp.scrs <- cbind(spp.scrs, Species = rownames(spp.scrs)) %>% 
  cbind(spp.pval) %>% 
  cbind(r2 = en$vectors$r) %>% 
  filter(p.value < 0.05, r2 > 0.5) %>% 
  left_join(bin_taxonomy)

scrs <- as.data.frame(scores(NMDS, display = "sites")) 

loadings <- as.data.frame(scores(NMDS, display = "species")) %>% rownames_to_column("features")

scaling_factor <- 50

(nmds_heterotroph_ko_with_env <- ggplot(scrs) +
  coord_fixed() + ## need aspect ratio of 1!
  geom_segment(data = spp.scrs,
               aes(x = 0, xend = NMDS1 * scaling_factor, y = 0, yend = NMDS2 * scaling_factor),
               arrow = arrow(length = unit(0.25, "cm")), colour = "grey") +
  ggrepel::geom_label_repel(data = spp.scrs, aes(x = NMDS1*scaling_factor, y = NMDS2*scaling_factor,  label = taxonomy),
            size = 2,
            alpha =0.6, max.overlaps = 100) +
  geom_point(mapping = aes(x = NMDS1, y = NMDS2)) +
  # ggrepel::geom_label_repel(mapping = aes(x = NMDS1, y = NMDS2, label = genus),
  #                           color = "red",
  #                           size = 1, alpha = 0.7) +
  labs(x = "NMDS1", y = "NMDS2") +
  #geom_point(data = loadings, aes(x = NMDS1, y = NMDS2, ko = ko),size = 0.1, alpha = 0.4) +
  #annotate(geom = 'text', label = paste0("Stress: ", round(drep_nmds$stress,2)), x = -Inf, y = -Inf, hjust = -0.3, vjust = -0.6) +
  theme_bw() ) 

```

#### With envfit microbes @ family level

```{r}
no_ti <- metadata %>% 
  filter(gut_section != "Terminal ileum")

veg_abund_no_TI <- vegan_abund[no_ti$import_id,]

CCA_res <- read_rds("results/CCA_multiomics_community_metabolite_simplified.rds") 

NMDS <- vegan::metaMDS(vegdist(vegan_abund[rownames(vegan_abund_family),],method = "bray"))

en <- envfit(NMDS, vegan_abund_family, permutations = 100, na.rm = TRUE)

env_res_df <- env_res$vectors$arrows



NMDS_points <- NMDS$points %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata)

#tax_fit <- envfit(NMDS, vegan_abund, permutations = 100, na.rm = TRUE)

colors = c("#9a5ebd","#2083e6","#2083e6", "#f09f26","#f09f26")
colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

(NMDS_points %>% 
  mutate(type_and_days = paste0(transplant_type, "_", days_after_transplant),
         location_type_and_days = paste0(gut_section, "_", type_and_days)) %>% 
  ggplot(aes(MDS1, MDS2, color = plot_categories, shape = gut_section)) +
  scale_alpha_discrete(range = c(0.4,1)) +
  geom_point(aes(group = plot_categories)) +
  ggforce::geom_mark_ellipse(aes(group = days_after_transplant, label = str_glue("{days_after_transplant} days")),label.fontsize = 10) +
  scale_color_manual(values = colors) +
  coord_fixed(xlim = c(-150, 100), ylim = c(-100, 60)) +
  labs(color = NULL, shape = NULL) +
  theme_bw() ) #%>% plotly::ggplotly()

bin_taxonomy <- rel.abund %>% 
  dplyr::select(taxonomy, Species = "taxonomy_id")


spp.pval <- en$vectors$pvals %>% data.frame(p.value = .)
spp.scrs <- as.data.frame(scores(en, display = "vectors")) 
spp.scrs <- cbind(spp.scrs, Species = rownames(spp.scrs)) %>% 
  cbind(spp.pval) %>% 
  cbind(r2 = en$vectors$r) %>% 
  filter(p.value < 0.05, r2 > 0.5) %>% rownames_to_column("taxonomy") #%>% 
  #left_join(bin_taxonomy)

scrs <- as.data.frame(scores(NMDS, display = "sites")) %>% rownames_to_column("sample") %>% left_join(metadata)

loadings <- as.data.frame(scores(NMDS, display = "species")) %>% rownames_to_column("features")

scaling_factor <- 1

(nmds_heterotroph_ko_with_env <- ggplot(scrs) +
  coord_fixed() + ## need aspect ratio of 1!
  geom_segment(data = spp.scrs,
               aes(x = 0, xend = NMDS1 * scaling_factor, y = 0, yend = NMDS2 * scaling_factor),
               arrow = arrow(length = unit(0.25, "cm")), colour = "grey") +
  ggrepel::geom_label_repel(data = spp.scrs, aes(x = NMDS1*scaling_factor, y = NMDS2*scaling_factor,  label = taxonomy),
            size = 2,
            alpha =0.6, max.overlaps = 100) +
  geom_point(mapping = aes(x = NMDS1, y = NMDS2, color = plot_categories)) +
    scale_color_manual(values = colors) + 
  # ggrepel::geom_label_repel(mapping = aes(x = NMDS1, y = NMDS2, label = genus),
  #                           color = "red",
  #                           size = 1, alpha = 0.7) +
  labs(x = "NMDS1", y = "NMDS2") +
  #geom_point(data = loadings, aes(x = NMDS1, y = NMDS2, ko = ko),size = 0.1, alpha = 0.4) +
  #annotate(geom = 'text', label = paste0("Stress: ", round(drep_nmds$stress,2)), x = -Inf, y = -Inf, hjust = -0.3, vjust = -0.6) +
  theme_bw() ) 

```


## Community similarity

```{r}
library(ggtree)

metadata_for_tree <- metadata %>% 
  dplyr::rename(label = "import_id")

sample_tree <- vegan_abund %>% 
  as.data.frame() %>% 
  rownames_to_column("import_id") %>% 
  left_join(metadata %>% dplyr::select(plot_categories, import_id, gut_section)) %>% 
  group_by(plot_categories, import_id) %>% 
  #mutate(plot_categories = str_glue("{gut_section}-{plot_categories}{row_number()}")) %>% 
  #column_to_rownames("plot_categories") %>% 
  column_to_rownames("import_id") %>% 
  dplyr::select(where(is.numeric)) %>% 
  vegdist(method = "robust.aitchison") %>% 
  hclust(method = "average")

colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")

ggtree::ggtree(sample_tree,layout='circular') %<+% (metadata %>% relocate(import_id)) +
  geom_tiplab(aes(label = str_glue("{gut_section} {if_else(replicate != 'other', replicate, '')}"))) + 
  expand_limits(x = 75) +
  geom_tippoint(aes(color = plot_categories)) +
  scale_color_manual(values = colors) + 
  labs(color = "Treatment group")

ggsave("results/community_distance_tree_circular.png", width = 4, height = 3, scale = 2.5)
ggsave("results/community_distance_tree_circular.pdf", width = 4, height = 3, scale = 2.5)


ggtree::ggtree(sample_tree) %<+% (metadata %>% relocate(import_id)) +
  geom_tiplab(aes(label = str_glue("{gut_section} {if_else(replicate != 'other', replicate, '')}"))) + 
  expand_limits(x = 75) +
  geom_tippoint(aes(color = plot_categories)) +
  scale_color_manual(values = colors) + 
  labs(color = "Treatment group")

ggsave("results/community_distance_tree_rect.png", width = 3, height = 4, scale = 2.5)
ggsave("results/community_distance_tree_rect.pdf", width = 3, height = 4, scale = 2.5)

```




## Differential abundance testing with ALDEX2

With full glm model
```{r}

aldex_abund <- rel.abund %>% 
  dplyr::select(taxonomy_id, any_of(metadata$import_id)) %>% 
  pivot_longer(-taxonomy_id, names_to = "sample", values_to = "abund") %>% 
  filter(abund > 0) %>% 
  group_by(sample) %>% 
  mutate(abund = round(abund / min(abund))) %>% 
  ungroup() %>% 
  pivot_wider(names_from = sample, values_from = abund,values_fill = 0) %>% 
  column_to_rownames("taxonomy_id")

mm <- model.matrix(~ transplant_type + gut_section + days_after_transplant, metadata %>% column_to_rownames("import_id") %>% .[colnames(aldex_abund),] %>% mutate(transplant_type = relevel(factor(transplant_type), "Syngeneic")))

mm_W_more_annot <- mm %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  left_join(metadata)

clr <- aldex.clr(aldex_abund,mm,mc.samples = 1000)

glm.test <- aldex.glm(clr, mm)

glm_res_w_tax <- glm.test %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(tax_data) %>% 
  relocate(taxonomy, taxonomy_id) %>%
  write_tsv("results/taxonomy/aldex_binTax_glm_results.tsv")

glm_res_w_tax <- read_tsv("results/taxonomy/aldex_binTax_glm_results.tsv")

glm.effect <- aldex.glm.effect(clr)
write_rds(glm.effect,"results/taxonomy/aldex_tax_glm.effect.rds")
#glm.effect <- read_rds("results/taxonomy/aldex_tax_glm.effect.rds")
```

```{r}
allogenic_effect <- glm.effect$transplant_typeAllogeneic %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(tax_data %>% mutate(taxonomy_id = as.character(taxonomy_id))) %>% 
  relocate(taxonomy,taxonomy_id)

cecum_effect <- glm.effect$gut_section.Q %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(tax_data %>% mutate(taxonomy_id = as.character(taxonomy_id))) %>% 
  relocate(taxonomy,taxonomy_id)

time_effect <- glm.effect$days_after_transplant %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(tax_data %>% mutate(taxonomy_id = as.character(taxonomy_id))) %>% 
  relocate(taxonomy,taxonomy_id)

```


```{r}
glm_res_w_tax %>% 
  separate(taxonomy, c("root", "kingdom", "phylum", "class", "order", "family", "genus", "species"),remove = FALSE,sep = ";[a-z]__") %>% 
  filter(`days_after_transplant.L:pval.holm` < 0.05 | `transplant_typeAllogeneic:pval.holm` < 0.05) %>% 
  mutate(significant_effect = case_when(`transplant_typeAllogeneic:pval.holm` < 0.05 ~ "Treatment",
                   `days_after_transplant.L:pval.holm` < 0.05 ~ "Time",
                   `transplant_typeAllogeneic:pval.holm` < 0.05 & `days_after_transplant.L:pval.holm` < 0.05 ~ "Both")) %>% 
  ggplot(aes(`transplant_typeAllogeneic:Est`,`days_after_transplant.L:Est`, color = class ,label = genus,shape = significant_effect)) + 
  geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +
  geom_point() + theme_bw() +
  ggrepel::geom_label_repel(max.overlaps = Inf,size = 2) +
  
  labs(y = "Day 7  ⬅ **Days after transplant** ➡  Day 21",
       x = "Syngeneic  ⬅ **Transplant type** ➡   Allogeneic",
       color = "Class",
       shape = "Significant\neffect") +
  theme(axis.title.x =  element_markdown(),
        axis.title.y = element_markdown())

ggsave("results/differential_abundance.png", width = 5, height = 3, scale = 1.25)  

```


#### With modified full glm model
```{r}
aldex_metadata <- metadata %>% 
  column_to_rownames("import_id") %>% 
  .[colnames(aldex_abund),] %>% 
  filter(transplant_type %in% c("Syngeneic", "Allogeneic")) %>% 
  mutate(transplant_type = relevel(factor(transplant_type), "Syngeneic"),
         gut_section = factor(gut_section, ordered = FALSE) %>% relevel("Transverse Colon"), 
         days_after_transplant = factor(days_after_transplant,ordered = FALSE) %>% relevel("7"))

aldex_abund <- rel.abund %>% 
  dplyr::select(taxonomy_id, any_of(metadata$import_id)) %>% 
  pivot_longer(-taxonomy_id, names_to = "sample", values_to = "abund") %>% 
  filter(abund > 0) %>% 
  group_by(sample) %>% 
  mutate(abund = round(abund / min(abund))) %>% 
  ungroup() %>% 
  pivot_wider(names_from = sample, values_from = abund,values_fill = 0) %>% 
  column_to_rownames("taxonomy_id") %>% 
  .[,rownames(aldex_metadata)]

mm <- model.matrix(~ transplant_type * days_after_transplant * gut_section, aldex_metadata)

mm_W_more_annot <- mm %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  left_join(metadata)

clr_full_int <- aldex.clr(aldex_abund,mm,mc.samples = 500) %>% 
  write_rds("results/aldex_glm_clr_full_int.rds")

aldex_glm_test_full_int <- aldex.glm(clr_full_int, mm) %>% 
  write_rds("results/aldex_glm_test_full_int.rds")

aldex_glm_test_full_int_w_tax <- aldex_glm_test_full_int %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(tax_data) %>% 
  relocate(taxonomy, taxonomy_id) %>%
  write_tsv("results/taxonomy/aldex_glm_test_full_int_w_tax.tsv")

aldex_glm_test_full_int_w_tax_long <- aldex_glm_test_full_int_w_tax %>% 
  pivot_longer(-c(taxonomy, taxonomy_id), names_to = "parameter", values_to = "value") %>% 
  mutate(parameter_clean = str_remove(parameter, ":[pval.padj,t.val,SE,Est]*$"),
         parameter_type = str_remove_all(parameter, ".*:"))

cleaned_wide <- aldex_glm_test_full_int_w_tax_long %>% 
  select(taxonomy, taxonomy_id, parameter_clean, parameter_type, value) %>% 
  pivot_wider(names_from = "parameter_type", values_from ="value")


allo_21_section_effect <- cleaned_wide %>% 
  filter(str_detect(parameter_clean, "transplant_typeAllogeneic:days_after_transplant21:gut_sectio")) %>% 
  arrange(pval) %>% 
  mutate(gut_section = str_remove(parameter_clean,"transplant_typeAllogeneic:days_after_transplant21:gut_section"))


glm_res_w_tax <- read_tsv("results/taxonomy/aldex_binTax_glm_results.tsv")

glm.effect <- aldex.glm.effect(clr)
write_rds(glm.effect,"results/taxonomy/aldex_tax_glm.effect.rds")
#glm.effect <- read_rds("results/taxonomy/aldex_tax_glm.effect.rds")
```


#### With full model @ genus level

With full glm model
```{r}

genus_abund <- rel.abund %>% 
  group_by(Domain, Phylum, Class, Order, Family, Genus) %>% 
  summarise(across(any_of(metadata$import_id), sum)) %>% 
  mutate(taxonomy_id = str_glue("d__{Domain};p__{Phylum};c__{Class};o__{Order};f__{Family};g__{Genus}")) %>% 
  relocate(taxonomy_id)

genus_tax <- genus_abund %>% select(taxonomy_id, Domain, Phylum, Class, Order, Family, Genus, -any_of(metadata$import_id))

aldex_abund <- genus_abund %>% 
  ungroup() %>% 
  dplyr::select(taxonomy_id, any_of(metadata$import_id)) %>% 
  pivot_longer(-taxonomy_id, names_to = "sample", values_to = "abund") %>% 
  filter(abund > 0) %>% 
  group_by(sample) %>% 
  mutate(abund = round(abund / min(abund))) %>% 
  ungroup() %>% 
  pivot_wider(names_from = sample, values_from = abund,values_fill = 0) %>% 
  column_to_rownames("taxonomy_id")

mm <- model.matrix(~ transplant_type * gut_section * days_after_transplant, metadata %>% column_to_rownames("import_id") %>% .[colnames(aldex_abund),] %>% mutate(transplant_type = relevel(factor(transplant_type), "Syngeneic")))

mm_W_more_annot <- mm %>% 
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  left_join(metadata)

clr_genus <- aldex.clr(aldex_abund,mm,mc.samples = 1000)

glm.test_genus <- aldex.glm(clr_genus, mm)

glm_res_w_tax_genus <- glm.test_genus %>% 
  rownames_to_column("taxonomy_id") %>% 
  left_join(genus_tax) %>% 
  relocate(taxonomy_id) %>%
  write_tsv("results/taxonomy/aldex_binTax_glm_results_genus.tsv")

glm_res_w_tax_genus <- read_tsv("results/taxonomy/aldex_binTax_glm_results_genus.tsv")

glm.effect_genus <- aldex.glm.effect(clr_genus)
write_rds(glm.effect_genus,"results/taxonomy/aldex_tax_glm.effect_genus.rds")
#glm.effect <- read_rds("results/taxonomy/aldex_tax_glm.effect.rds")
```


```{r}
glm_res_w_tax_genus %>% 
  filter(`days_after_transplant.L:pval.padj` < 0.00001 | `transplant_typeAllogeneic:pval.padj` < 0.00001) %>% 
  mutate(significant_effect = case_when(`transplant_typeAllogeneic:pval.padj` < 0.01 ~ "Treatment",
                   `days_after_transplant.L:pval.padj` < 0.01 ~ "Time",
                   `transplant_typeAllogeneic:pval.padj` < 0.01 & `days_after_transplant.L:pval.padj` < 0.01 ~ "Both")) %>% 
  ggplot(aes(`transplant_typeAllogeneic:Est`,`days_after_transplant.L:Est`, color = Class ,label = Genus,shape = significant_effect)) + 
  geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +
  geom_point() + theme_bw() +
  ggrepel::geom_label_repel(max.overlaps = Inf,size = 2) +
  labs(y = "Day 7  ⬅ **Days after transplant** ➡  Day 21",
       x = "Syngeneic  ⬅ **Transplant type** ➡   Allogeneic",
       color = "Class",
       shape = "Significant\neffect") +
  theme(axis.title.x =  element_markdown(),
        axis.title.y = element_markdown())

ggsave("results/differential_abundance_genus.png", width = 5, height = 3, scale = 1.75)  

```



# All sites
## Syngeneic vs allogeneic (all)
```{r}
# taxonomy <- read_qza("data/calcrete/silva_132_taxonomy.qza")$data %>% 
#   dplyr::select(otu = "Feature.ID", taxonomy = "Taxon")

#met_samples <- read_tsv("data/calcrete/metadata.tsv")

allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic")) %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic")) %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

otu_genus_summary_allo_syn_all <- otus %>% 
  mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c( "#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Syngeneic vs allogeneic: day 7
```{r}
# taxonomy <- read_qza("data/calcrete/silva_132_taxonomy.qza")$data %>% 
#   dplyr::select(otu = "Feature.ID", taxonomy = "Taxon")

#met_samples <- read_tsv("data/calcrete/metadata.tsv")

allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7) %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7) %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

otu_genus_summary_allo_syn_all <- otus %>% 
  mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c( "#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic_day7.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic_day7.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Syngeneic vs allogeneic: day 21
```{r}
# taxonomy <- read_qza("data/calcrete/silva_132_taxonomy.qza")$data %>% 
#   dplyr::select(otu = "Feature.ID", taxonomy = "Taxon")

#met_samples <- read_tsv("data/calcrete/metadata.tsv")

allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21) %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21) %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

otu_genus_summary_allo_syn_all <- otus %>% 
  mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c( "#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic_day21.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Allogeneic_vs_Syngeneic_heat_tree, "results/Allogeneic_vs_Syngeneic_day21.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Naive vs allogeneic (all)
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic")) %>%
  pull("import_id")
  
naive_samples <- metadata %>%
  filter(transplant_type %in% c("Naive")) %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,naive_samples)

sample_types <- c("Allogeneic", "Naive")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

# otu_genus_summary_allo_naive_all <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_naive = Allogeneic / Naive)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Allogeneic_vs_naive_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","green4"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Allogeneic_vs_naive_heat_tree, "results/naive_vs_Allogeneic.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Allogeneic_vs_naive_heat_tree, "results/naive_vs_Allogeneic.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Naice vs. Syngeneic (all)
```{r}
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic")) %>%
  pull("import_id")
  
naive_samples <- metadata %>%
  filter(transplant_type %in% c("Naive")) %>%
  pull("import_id")
  
samples_to_compare <- c(syngeneic_samples,naive_samples)

sample_types <- c("Syngeneic", "Naive")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

# otu_genus_summary_syn_naive_all <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_naive = Allogeneic / Naive)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Syngeneic_vs_naive_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#3279a8","grey80", "green4"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Syngeneic_vs_naive_heat_tree, "results/naive_vs_Syngeneic.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Syngeneic_vs_naive_heat_tree, "results/naive_vs_Syngeneic.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Syngeneic day 7 vs 21
```{r}
# taxonomy <- read_qza("data/calcrete/silva_132_taxonomy.qza")$data %>% 
#   dplyr::select(otu = "Feature.ID", taxonomy = "Taxon")

#met_samples <- read_tsv("data/calcrete/metadata.tsv")

day7_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7) %>%
  pull("import_id")
  
day21_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21) %>%
  pull("import_id")
  
samples_to_compare <- c(day7_samples,day21_samples)

sample_types <- c("7", "21")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(days_after_transplant %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(days_after_transplant) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(days_after_transplant, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = days_after_transplant, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

otu_genus_summary_allo_syn_all <- otus %>% 
  mutate(allo_to_syn = `7` / `21`)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(syn_day7_vs_day21_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c( "#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = syn_day7_vs_day21_heat_tree, "results/Syngeneic_day7_vs_day21_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = syn_day7_vs_day21_heat_tree, "results/Syngeneic_day7_vs_day21_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

## Allogeneic day 7 vs 21
```{r}
day7_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7) %>%
  pull("import_id")
  
day21_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21) %>%
  pull("import_id")
  
samples_to_compare <- c(day7_samples,day21_samples)

sample_types <- c("7", "21")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(days_after_transplant %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(days_after_transplant) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(days_after_transplant, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = days_after_transplant, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

otu_genus_summary_allo_syn_all <- otus %>% 
  mutate(allo_to_syn = `7` / `21`)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(allo_day7_vs_day21_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c( "#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = allo_day7_vs_day21_heat_tree, "results/Allogeneic_day7_vs_day21_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = allo_day7_vs_day21_heat_tree, "results/Allogeneic_day7_vs_day21_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Terminal Illeum
### Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(termIleum_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80", "#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


ggsave(plot = termIleum_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
ggsave(plot = termIleum_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

### Day 7 Terminal Illeum: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7,
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7,
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(termIleum_Day7_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = termIleum_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Day7_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = termIleum_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Day7_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


### Day 21 Terminal Illeum: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21,
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21,
         gut_section == "Terminal ileum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
  pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
  mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
  group_by(taxonomy,type) %>% 
  summarise(abund = sum(abund)) %>% 
  pivot_wider(names_from = type,values_from = abund) %>% 
  mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(termIleum_Day21_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = termIleum_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Day21_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = termIleum_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/termIleum_Day21_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

## Cecum
### Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         gut_section == "Cecum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         gut_section == "Cecum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Cecum_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Cecum_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Cecum_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

### Day 7 Cecum: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7,
         gut_section == "Cecum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7,
         gut_section == "Cecum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Cecum_Day7_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Cecum_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Day7_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Cecum_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Day7_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


### Day 21 Cecum: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21,
         gut_section == "Cecum") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21,
         gut_section == "Cecum") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Cecum_Day21_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Cecum_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Day21_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Cecum_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Cecum_Day21_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


## Transverse Colon
### Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Transverse_Colon_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Transverse_Colon_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Transverse_Colon_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

### Day 7 Transverse Colon: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7,
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7,
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Transverse_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Transverse_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Transverse_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


### Day 21 Transverse Colon: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21,
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21,
         gut_section == "Transverse Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Transverse_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Transverse_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Transverse_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Transverse_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

## Descending Colon
### Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Descending_Colon_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Descending_Colon_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Descending_Colon_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

### Day 7 Descending Colon: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 7,
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 7,
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Descending_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Descending_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Descending_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Day7_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```


### Day 21 Transverse Colon: Syngeneic vs allogeneic
```{r}
allogeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Allogeneic"),
         days_after_transplant == 21,
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
syngeneic_samples <- metadata %>%
  filter(transplant_type %in% c("Syngeneic"),
         days_after_transplant == 21,
         gut_section == "Descending Colon") %>%
  pull("import_id")
  
samples_to_compare <- c(allogeneic_samples,syngeneic_samples)

sample_types <- c("Allogeneic", "Syngeneic")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(transplant_type %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(transplant_type) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(transplant_type, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = transplant_type, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

# otu_genus_summary_allo_syn_termIl_day21 <- otus %>% 
#   pivot_longer(-taxonomy,names_to = "type", values_to = "abund") %>% 
#   mutate(taxonomy = str_remove(taxonomy,";s__.*")) %>% 
#   group_by(taxonomy,type) %>% 
#   summarise(abund = sum(abund)) %>% 
#   pivot_wider(names_from = type,values_from = abund) %>% 
#   mutate(allo_to_syn = Allogeneic / Syngeneic)

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 

#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh

obj <- filter_obs(obj,"tax_abund",tab)

keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)

obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))

#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)


obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)

(Descending_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree <- obj %>%
      heat_tree(
                 node_size = Combined,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 #node_color_range = viridis::viridis(3),
                 node_color_range = c("#ffc559","grey80","#3279a8"),
                 node_color_trans = "linear",
                 node_color_interval = c(3, -3),
                 edge_color_interval = c(3, -3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 80,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))


  ggsave(plot = Descending_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree.pdf", device = cairo_pdf, width = 6, height = 6, dpi = 600)
  ggsave(plot = Descending_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree, "results/Descending_Colon_Day21_Allogeneic_vs_Syngeneic_heat_tree.png", type = "cairo", width = 6, height = 6, dpi = 600)
```

## Multi-site comparisons
### All samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)
sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_tree.png", width = 6, height = 6)
```


### All samples except terminal illeum
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon") %>%
  pull("import_id")

samples_to_compare <- c(cecum_samples,transColon_samples,descColon_samples)
sample_types <- c("Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_tree_noTI.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_tree_noTI.png", width = 6, height = 6)
```

### All Syngeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         transplant_type =="Syngeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")


otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_tree.png", width = 6, height = 6)
```

### Syngeneic samples, all except terminal illeum
```{r}
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         transplant_type =="Syngeneic") %>%
  pull("import_id")

samples_to_compare <- c(cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Cecum", "Transverse Colon", "Descending Colon")


otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_tree_noTI.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_tree_noTI.png", width = 6, height = 6)
```



### 7-day post transplant Syngeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         days_after_transplant == 7,
         transplant_type =="Syngeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         days_after_transplant == 7,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         days_after_transplant == 7,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         days_after_transplant == 7,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_Day7_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_Day7_tree.png", width = 6, height = 6)
```

### 21-day post transplant Syngeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         days_after_transplant == 21,
         transplant_type =="Syngeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         days_after_transplant == 21,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         days_after_transplant == 21,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         days_after_transplant == 21,
         transplant_type =="Syngeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)
sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_Day21_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Syngeneic_Day21_tree.png", width = 6, height = 6)
```



### All Allogeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         transplant_type =="Allogeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_tree.png", width = 6, height = 6)
```
### Allogeneic samples, all except TI
```{r}
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         transplant_type =="Allogeneic") %>%
  pull("import_id")

samples_to_compare <- c(cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_tree_noTI.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_tree_noTI.png", width = 6, height = 6)
```


### 7-day post transplant Allogeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         days_after_transplant == 7,
         transplant_type =="Allogeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         days_after_transplant == 7,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         days_after_transplant == 7,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         days_after_transplant == 7,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)

sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 


#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_Day7_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_Day7_tree.png", width = 6, height = 6)
```

### 21-day post transplant Allogeneic samples
```{r}
termIleum_samples <- metadata %>%
  filter(gut_section == "Terminal ileum",
         days_after_transplant == 21,
         transplant_type =="Allogeneic") %>%
  pull("import_id")
  
cecum_samples <- metadata %>%
  filter(gut_section == "Cecum",
         days_after_transplant == 21,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

transColon_samples <- metadata %>%
  filter(gut_section == "Transverse Colon",
         days_after_transplant == 21,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

descColon_samples <- metadata %>%
  filter(gut_section == "Descending Colon",
         days_after_transplant == 21,
         transplant_type =="Allogeneic") %>%
  pull("import_id")

samples_to_compare <- c(termIleum_samples,cecum_samples,transColon_samples,descColon_samples)
sample_types <- c("Terminal ileum", "Cecum", "Transverse Colon", "Descending Colon")

otus <- rel.abund %>% 
  pivot_longer(one_of(metadata$sample), names_to = "sample", values_to = "abund") %>% 
  left_join(metadata) %>% 
  filter(gut_section %in% sample_types,
         sample %in% samples_to_compare) %>% 
  group_by(gut_section) %>% 
  mutate(group_sum = sum(abund)) %>% 
  group_by(gut_section, taxonomy) %>% 
  summarise(abund = sum(abund / group_sum)) %>% 
  pivot_wider(names_from = gut_section, values_from = abund) %>% 
  ungroup() %>% 
  mutate(Combined = rowSums(across(-taxonomy)) / length(sample_types)) #For overall structure, equally weighted relative abundance between the sample types 

#Read parse the data for metacodeR and group all taxa lower than genus into their genus supertaxa
obj <- parse_tax_data(otus, class_cols = "taxonomy", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$") %>% 
        metacoder::filter_taxa(taxon_ranks == "g", 
                               supertaxa = TRUE,
                               reassign_obs = TRUE, 
                               subtaxa = FALSE,
                               drop_obs = FALSE)
#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data",
                                       cols = c(sample_types,"Combined")) 
#Only plot organisms with abundance greater than threshold 
thresh <- 0.01 #Rel.abund threshold to filter for plottting
tab <- obj$data$tax_abund %>% dplyr::select(-taxon_id) %>% rowSums() > thresh
obj <- filter_obs(obj,"tax_abund",tab)
keep <- obj$data$tax_data$taxon_id %in% obj$data$tax_abund$taxon_id
obj <- filter_obs(obj,"tax_data", keep)
obj <- metacoder::filter_taxa(obj,unique(obj$data$tax_abund$taxon_id))
#adds comparison data about type to the data
obj$data$diff_table <- compare_groups(obj, data = "tax_abund",
                                      cols = sample_types,
                                      groups = sample_types)
obj$data$tax_data <- NULL #otherwise plot gets confused looking for concrete (rel.abund)
colors <- c("#ef8a62","grey70", "#67a9cf")
(heat_tree <- obj %>%
heat_tree_matrix(data = "diff_table",
                 node_size = Combined,
                 key_size = 0.7,
                 node_label = taxon_names,
                 node_color = log2_median_ratio,
                 # node_color_range = viridis::viridis(3),
                 # row_label_color = viridis::viridis(3)[3],
                 # col_label_color = viridis::viridis(3)[1],
                 node_color_range = colors,
                 row_label_color = colors[3],
                 col_label_color = colors[1],
                 node_color_trans = "linear",
                 node_color_interval = c(-3, 3),
                 edge_color_interval = c(-3, 3),
                 node_size_axis_label = "Relative abundance",
                 node_color_axis_label = "Log2 ratio median proportions",
                 node_size_range = c(0.0005,0.05),
                 node_label_size_range = c(.015,.02),
                 node_label_max = 100,
                 initial_layout = "reingold-tilford", layout = "davidson-harel",
                 overlap_avoidance = 1.2))
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_Day21_tree.pdf", width = 6, height = 6)
  ggsave(plot = heat_tree, "results/gut_section_Allogeneic_Day21_tree.png", width = 6, height = 6)
```


# Plot abundance of particular taxa

## Lactobacillus

Nov 8, 2022 request:  relative abundance of Lactobacillus (in a bar graph or something) in the descending colon, d21 post-bmt, syn vs allo
```{r}

d21_desc_samples <- metadata %>% 
  filter(gut_section =="Descending Colon", 
         days_after_transplant == 21, 
         transplant_type %in% c("Allogeneic", "Syngeneic"))

(lacto_abund_plt <- rel_abund_long %>% 
  filter(Genus =="Lactobacillus",
         sample %in% d21_desc_samples$sample) %>% 
  left_join(d21_desc_samples, by ="sample") %>% 
  group_by(sample, gut_section, days_after_transplant, transplant_type) %>% 
  summarise(percent_abund = sum(rel_abund) * 100) %>% 
  ggplot(aes(transplant_type, percent_abund)) +
  geom_boxplot() +
  geom_jitter(width = 0.2) +
  theme_bw() +
  labs(y = "Percent abundance",
       x = NULL, title = "*Lactobacillus*") +
  theme(plot.title = ggtext::element_markdown())

)

ggsave("results/Lactobacillus_d21_DescendingColon.pdf", width = 4, height = 3, dpi = 600, scale = 1)
ggsave("results/Lactobacillus_d21_DescendingColon.png", width = 4, height = 3, dpi = 600, scale = 1)



lacto_abund_plt$data %>% 
  ggboxplot(x = "transplant_type", y = "percent_abund",
          color = "transplant_type", palette = "jco",
          add = "jitter") +
  stat_compare_means(aes(group = transplant_type),method = "t.test") +
    labs(x = NULL, y = "Percent abundance", color = NULL) +
  labs(y = "Percent abundance",
       x = NULL, title = "*Lactobacillus*") +
  theme(plot.title = ggtext::element_markdown())

ggsave("results/Lactobacillus_d21_DescendingColon_stats.pdf", width = 4, height = 3, dpi = 600, scale = 1.2)
ggsave("results/Lactobacillus_d21_DescendingColon_stats.png", width = 4, height = 3, dpi = 600, scale = 1.2)
```


# E. coli
```{r}

taxa = "Akkermansia"
level = "Genus"

plot_taxa <- function(taxa, level){
  all_tax_levels <- factor(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id"), levels = rev(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id")), ordered = TRUE)

  grouping_levels <- all_tax_levels[all_tax_levels >= level]
  
  colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")
  
  col_name <- sym(level)
  abund_df <- rel_abund_long %>% 
    group_by(sample) %>% 
    mutate(rel_abund = rel_abund / sum(rel_abund)) %>% 
    group_by(across(c(sample, grouping_levels))) %>% 
    summarise(rel_abund = sum(rel_abund),.groups = "drop") %>% 
    left_join(metadata) %>% 
    filter(!!col_name == taxa) %>% 
    ggplot(aes(1, rel_abund, fill = plot_categories)) +
    geom_boxplot() +
    theme_bw() +
    theme(axis.ticks.x = element_blank(),
          axis.text.x = element_blank()) + 
    scale_fill_manual(values = colors) +
    facet_grid(~gut_section, scales = "free_x", labeller = labeller(gut_section = label_wrap_gen(5))) + 
    scale_y_log10() +
    labs(x = NULL,
         y = "% abundance (log10)", 
         fill = NULL,
         title = str_glue("{taxa} ({tolower(level)})"))
  
  return(abund_df)
}


taxa = "Escherichia"
level = "Genus"


plot_taxa_w_stats <- function(taxa, level){
  all_tax_levels <- factor(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id"), levels = rev(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id")), ordered = TRUE)

  grouping_levels <- all_tax_levels[all_tax_levels >= level]
  
  colors_df = data.frame(plot_categories = factor(x = c("Naive","Syngeneic; day 7","Syngeneic; day 21", "Allogeneic; day 7", "Allogeneic; day 21"),
                                                  levels = c("Naive","Syngeneic; day 7","Syngeneic; day 21", "Allogeneic; day 7", "Allogeneic; day 21"),
                                                  ordered = TRUE),
             colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26"))
  
  
  col_name <- sym(level)
  abund_df <- rel_abund_long %>% 
    group_by(sample) %>% 
    mutate(rel_abund = ((rel_abund / sum(rel_abund))+.00001)*100) %>% 
    group_by(across(c(sample, grouping_levels))) %>% 
    summarise(rel_abund = sum(rel_abund),.groups = "drop") %>% 
    left_join(metadata) %>% 
    filter(!!col_name == taxa,
           rel_abund > 0) %>% 
    left_join(colors_df) %>%
    ggboxplot(x = "plot_categories", y = "rel_abund",
          fill = "colors", palette = "jco") +
    #scale_color_manual(values = colors) +
    ggpubr::stat_compare_means(comparisons = my_comparisons,method = "t.test",hide.ns = TRUE,label = "p.signif",na.rm = TRUE,vjust = 0.8) +
    #geom_pwc(,method = "t.test",hide.ns = TRUE) +
    
    # ggplot(aes(1, rel_abund, fill = plot_categories)) +
    # geom_boxplot() +
    theme_bw() +
    theme(axis.ticks.x = element_blank(),
          axis.text.x = element_blank()) + 
    #scale_fill_manual(values = colors) +
    scale_fill_identity(guide = 'legend',
                      labels = colors_df$plot_categories,
                      breaks =  colors_df$colors) +
    facet_grid(~gut_section, scales = "free_x", labeller = labeller(gut_section = label_wrap_gen(5))) + 
    scale_y_log10(labels = scales::comma) +
    labs(x = NULL,
         y = "% abundance (log10)", 
         fill = NULL,
         title = str_glue("{taxa} ({tolower(level)})"))
  
  return(abund_df)
}


plot_taxa_line <- function(taxa, level){
  all_tax_levels <- factor(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id"), levels = rev(c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species", "taxonomy_id")), ordered = TRUE)

  grouping_levels <- all_tax_levels[all_tax_levels >= level]
  
  colors = c("#9a5ebd","#9dbfe0","#2083e6", "#dec8a6","#f09f26")
  
  col_name <- sym(level)
  abund_df <- rel_abund_long %>% 
    group_by(sample) %>% 
    mutate(rel_abund = rel_abund / sum(rel_abund)) %>% 
    group_by(across(c(sample, grouping_levels))) %>% 
    summarise(rel_abund = sum(rel_abund),.groups = "drop") %>% 
    left_join(metadata) %>% 
    filter(!!col_name == taxa) %>% 
    ggplot(aes(as.numeric(gut_section), rel_abund, color = plot_categories)) +
    geom_point() +
    geom_smooth(se = FALSE) +
    theme_bw() +
    theme(axis.ticks.x = element_blank(),
          axis.text.x = element_blank()) + 
    scale_color_manual(values = colors) +
    #facet_grid(~day, scales = "free_x", labeller = labeller(gut_section = label_wrap_gen(5))) + 
    scale_y_log10() +
    labs(x = NULL,
         y = "% abundance (log10)", 
         fill = NULL,
         title = str_glue("{taxa} ({tolower(level)})"))
  
  return(abund_df)
}

plot_taxa_line(taxa = "Eubacterium_R", level = "Genus")
```


```{r}
tax_to_plot <- "Escherichia"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)

```

```{r}
tax_to_plot <- "Akkermansia"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Lactobacillus"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Clostridia"
tax_level <- "Class"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Lachnospirales"
tax_level <- "Order"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Oscillospirales"
tax_level <- "Order"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Enterococcus"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```



```{r}
tax_to_plot <- "Emergencia"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Ruminococcus"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

```{r}
tax_to_plot <- "Muribaculaceae"
tax_level <- "Family"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```



```{r}
tax_to_plot <- "Coproplasma"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```


# Naive abund
```{r}
tax_to_plot <- "Paramuribaculum"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```


# Descending colon abund
```{r}
tax_to_plot <- "Ruminococcus_E"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)

#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```


## Terminal Ileum abund
```{r}
tax_to_plot <- "Dwaynesavagella"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```



### Day 21 abund

Often found in inflammed conditions, highliy immunologic: https://www.frontiersin.org/articles/10.3389/fcimb.2015.00084/full
```{r}
tax_to_plot <- "Erysipelotrichaceae"
tax_level <- "Family"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```

## Day 7 abund

Involved in bile acid transformations: https://www.biorxiv.org/content/10.1101/2022.06.27.497673v1.full
```{r}
tax_to_plot <- "Turicibacter"
tax_level <- "Genus"

plot_taxa(tax_to_plot,tax_level)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}.pdf"), width = 2, height = 1, scale = 3)


#Plot with stats
(assign(str_glue("{tax_to_plot}_stats_plot"), plot_taxa_w_stats(tax_to_plot,tax_level)))
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.png"), width = 2, height = 1, scale = 3, dpi = 300)
ggsave(str_glue("results/individual_taxa_plots/{tax_to_plot}_stats.pdf"), width = 2, height = 1, scale = 3)
```


## Bifidobacteria
```{r}
(bifido_abund_plt <- rel_abund_long %>% 
  filter(Genus =="Bifidobacterium",
         ) %>% 
  left_join(d21_desc_samples, by ="sample") %>% 
  group_by(sample, gut_section, days_after_transplant, transplant_type) %>% 
  summarise(percent_abund = sum(rel_abund) * 100) %>% 
  ggplot(aes(transplant_type, percent_abund)) +
  geom_boxplot() +
  geom_jitter(width = 0.2) +
  theme_bw() +
  labs(y = "Percent abundance",
       x = NULL, title = "*Lactobacillus*") +
  theme(plot.title = ggtext::element_markdown())

)

ggsave("results/Lactobacillus_d21_DescendingColon.pdf", width = 4, height = 3, dpi = 600, scale = 1)
ggsave("results/Lactobacillus_d21_DescendingColon.png", width = 4, height = 3, dpi = 600, scale = 1)

```

look in bracken data instead since it looks like we have no bifido bins
```{r}
bifido_bracken <- rel_abund_long_bracken %>% 
  filter(Genus == "g__Bifidobacterium",
         rel_abund > 1e-05) %>% 
  left_join(metadata)

bifido_bracken %>% 
  ggplot(aes(gut_section,rel_abund )) +
  geom_boxplot() +
  geom_jitter(color = "blue", alpha = 0.4) +
  facet_grid(str_wrap(str_remove(Species,"^s__"),width = 18,whitespace_only = FALSE) ~ plot_categories ) +
  theme_bw() +
  #coord_flip() 
  scale_x_discrete(guide = guide_axis(angle = -65)) 

ggsave("results/bifido.pdf", width = 3, height = 6, scale = 3)

```


#### Combined plot with significance for paper 
```{r}
library(patchwork)


(Escherichia_stats_plot + Enterococcus_stats_plot) / (Akkermansia_stats_plot + Ruminococcus_stats_plot) + plot_layout(guides = "collect")

ggsave("results/individual_taxa_plots/combined_taxa_w_stats.png", width = 3, height = 1.5, scale = 3.5)
ggsave("results/individual_taxa_plots/combined_taxa_w_stats.pdf", width = 4, height = 1, scale = 3)
ggsave("results/individual_taxa_plots/combined_taxa_w_stats.eps", width = 4, height = 1, scale = 3)
ggsave("results/individual_taxa_plots/combined_taxa_w_stats.svg", width = 4, height = 1, scale = 3)
```




# Random Forest community differences

```{r}
library(tidymodels)

percent_abund_wide_w_met <- percent_abund %>% 
  #select(genome, percent_abund, sample, transplant_type, days_after_transplant, gut_section, replicate) %>% 
  select(genome, percent_abund, sample, transplant_type = "plot_categories", days_after_transplant, gut_section, replicate) %>% 
  pivot_wider(names_from = genome, values_from = percent_abund)

classified_metabolites <- metabolite_info %>% 
  filter(!is.na(sub_class),!is.na(`Feature Name or Formula`))

percent_abund_wide_w_met <- metab_combined %>% 
  t() %>% 
  as.data.frame() %>% 
  select(any_of(classified_metabolites$`Feature Name or Formula`)) %>% 
  rownames_to_column("sample") %>% 
  left_join(metadata %>% select(sample, transplant_type = "plot_categories", days_after_transplant, gut_section, replicate))

community_boot <- bootstraps(percent_abund_wide_w_met)

preprocess_plan <- recipe(transplant_type ~ ., data = percent_abund_wide_w_met) %>% 
  update_role(days_after_transplant,new_role = "Id") %>% 
  update_role(sample,new_role = "Id") %>% 
  update_role(gut_section,new_role = "Id") %>% 
  update_role(replicate,new_role = "Id") %>% 
  step_zv(all_predictors()) %>% 
  step_normalize(all_predictors())

community_prep <- prep(preprocess_plan)
```

```{r}
rf_spec <- rand_forest(trees = 1000) %>% 
  set_mode("classification") %>% 
  set_engine("ranger")

community_rf_flow <- workflow() %>% 
  add_recipe(preprocess_plan) %>% 
  add_model(rf_spec)

```

```{r}
community_rf_res <- fit_resamples(
  community_rf_flow,
  resamples = community_boot,
  control = control_resamples(save_pred = TRUE, 
                              verbose = TRUE)
)
```

Collect the metrics
```{r}
community_rf_res %>% 
  collect_metrics()
```
Confusion matrix
```{r}
community_rf_res %>% 
  collect_predictions() %>% 
  conf_mat(transplant_type, .pred_class)
```
Positive predictive value
```{r}
community_rf_res %>% 
  collect_predictions() %>% 
  ppv(transplant_type, .pred_class)
```

```{r}
library(vip)

rf_w_vip <- rf_spec %>% 
  set_engine("ranger", importance = "permutation") %>% 
  fit(transplant_type ~ .,
      data = juice(community_prep) %>% select(-c(sample,days_after_transplant, gut_section, replicate))) #%>%
  #vip(geom = "point")
  
  
import <- rf_w_vip[["fit"]][["variable.importance"]] %>% 
  data.frame(importance = ., genome = names(.)) %>% 
  #left_join(percent_abund %>% ungroup() %>% select(genome, classification, Phylum, Class, Order, Family, Genus, Species) %>% distinct()) %>% 
  rename(`Feature Name or Formula` = "genome") %>% left_join(classified_metabolites) %>% 
  arrange(desc(importance)) %>% 
  write_tsv("results/random_forest_res.tsv")
  
```


```{r}

metab_combined_long %>% 
  dplyr::rename(import_id  = "sample_id") %>% 
  left_join(metab_metadata) %>% 
  filter(compound == "C23H44O4_Idx-10531") %>% 
  ggplot(aes(abund, plot_categories, )) +
  geom_point() +
  theme_bw()

top_20_impotance_bile_acids <-  import %>%
  filter(main_class == "Bile acids") %>% 
  slice_max(order_by = importance, n = 20)


metab_combined_long %>% 
  dplyr::rename(import_id  = "sample_id") %>% 
  left_join(metab_metadata) %>% 
  filter(compound %in% top_20_impotance_bile_acids$`Feature Name or Formula`) %>% 
  ggplot(aes(abund, plot_categories)) +
  geom_point() +
  theme_bw() +
  facet_grid(~compound,scales = "free_x")
```




